### DISCUSSION TITLE
How to test ratatui with insta snapshots?

### DISCUSSION THREAD

**Discussion by rcorre** [ratatui/ratatui#1381] [Q&A] [answered]
The [counter tutorial](https://ratatui.rs/tutorials/counter-app/basic-app/#testing-the-ui-output) shows how to write tests that use `assert_eq!(buffer, buffer)`, but it seems like constructing all those buffers by hand would be painstaking if you have many tests and a complex UI.

I thought I'd try [insta](https://insta.rs/docs/cli/), but `assert_snapshot` expects a `str`. Is there any way to convert a `Buffer` to a `String`, or some other way of using snaphot tests?

**Comment by rcorre** [comment]
I came up with something, I'm not sure it's the best approach, it doesn't handle styling, and it may not handle multi-character graphemes, but it works so far:

```
    #[test]
    fn render() {
        let app = App::default();
        let mut buf = Buffer::empty(Rect::new(0, 0, 50, 4));

        app.render(buf.area, &mut buf);

        let actual = buf
            .content
            .chunks(buf.area.width as usize)
            .map(|line| {
                line.iter()
                    .map(|cell| cell.symbol().to_string())
                    .collect::<Vec<_>>()
                    .join("")
            })
            .collect::<Vec<_>>()
            .join("\n");

        assert_snapshot!(actual);
    }
```

**Reply by kdheepak** [reply]
Possibly related: https://github.com/ratatui/ratatui/issues/1045

**Comment by drmorr0** [comment]
I'm doing this slightly differently: when you call `teminal.draw(|f| ...)` it returns a [CompletedFrame](https://docs.rs/ratatui/latest/ratatui/struct.CompletedFrame.html) which contains all the contents of the frame (including style information).  It doesn't implement Display, but it does implement Debug, which means you can do snapshot testing with insta and `assert_debug_snapshot`:

```
fn test_render_event_list(mut test_app: App) {
    let mut term = Terminal::new(TestBackend::new(80, 20)).unwrap();
    let cf = term.draw(|frame| view(&mut test_app, frame)).unwrap();
    assert_debug_snapshot!(cf);
}
```

I'm also not sure how useful/maintainable this will be in the long term, but for right now it's working OK.

**Comment by joshka** [comment]
> but it seems like constructing all those buffers by hand would be painstaking if you have many tests and a complex UI.

We got by in adding tests to all the widgets generally by making the debug output of the assertion copy-paste friendly. This meant that the test flow was:
1. write an intentionally broken test
2. look at the output of the test and confirm that it matches expectations
3. copy output to test

This worked surprisingly well in practice and can be seen as the `ratatui::widgets` module has about 98% coverage.

**Comment by orhun** [comment]
I tried adding a snapshot test for demo2 for fun: #1395

It worked better than I expected and it's a good demonstration of how this might work. Let me know what you think, maybe we could consider adding more snapshot tests.

**Reply by rcorre** [reply]
Oh, that's cool, I didn't know about `TestBackend`. That seems like the cleanest way to get a string out of the render result.

**Reply by joshka** [reply]
Probably a good idea would be to make this a website write up.

**Reply by orhun** [reply, marked as answer]
https://github.com/ratatui/ratatui-website/pull/739

**Comment by a-kenji** [comment]
Here examples on how I snapshot test [`tui-term`](https://github.com/a-kenji/tui-term):

- The test helper function for my specific needs: https://github.com/a-kenji/tui-term/blob/efa760551bc24a3fb55c677b5f2954be9af7efea/src/widget.rs#L308
- Loading in an simple ls output: https://github.com/a-kenji/tui-term/blob/efa760551bc24a3fb55c677b5f2954be9af7efea/src/widget.rs#L357
- Overlapping cursors with extra modifications from ratatui: https://github.com/a-kenji/tui-term/blob/efa760551bc24a3fb55c677b5f2954be9af7efea/src/widget.rs#L438
- Documentation on how to record snapshots and use insta: https://github.com/a-kenji/tui-term/blob/development/docs/CONTRIBUTING.md#how-to-record-script-tests

To support this use case I supported making the test buffer public: 
- https://github.com/ratatui/ratatui/pull/160
- https://github.com/ratatui/ratatui/pull/161
